tests/test_transf_comp.py

```diff
@@ -1,29 +1,33 @@
-import unittest
+import pytest
 import numpy as np
 from cpmpy import *
 from cpmpy.transformations.flatten_model import flatten_constraint
 from cpmpy.transformations.comparison import only_numexpr_equality
 from cpmpy.expressions.variables import _IntVarImpl, _BoolVarImpl # to reset counters
 
-class TestTransfComp(unittest.TestCase):
-    def setUp(self):
-        _IntVarImpl.counter = 0
-        _BoolVarImpl.counter = 0
+def _generate_inputs():
+    ivs = intvar(1,9, shape=3, name="x")
+    for case in [
+            (min(ivs) == 3, "[min(x[0],x[1],x[2]) == 3]"),
+            (min(ivs) > 3, "[(min(x[0],x[1],x[2])) == (IV0), IV0 > 3]"),
+            (min(ivs) <= 3, "[(min(x[0],x[1],x[2])) == (IV0), IV0 <= 3]"),
+            (3 != max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 != 3]"),
+            (3 > max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 < 3]"),
+            (3 <= max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 >= 3]"),
+            ]:
+        yield case
 
-    def test_only_numexpr_eq(self):
-        ivs = intvar(1,9, shape=3, name="ivs")
-        
-        cases = [(min(ivs) == 3, "[min(ivs[0],ivs[1],ivs[2]) == 3]"),
-                 (min(ivs) > 3, "[(min(ivs[0],ivs[1],ivs[2])) == (IV0), IV0 > 3]"),
-                 (min(ivs) <= 3, "[(min(ivs[0],ivs[1],ivs[2])) == (IV2), IV2 <= 3]"),
-                 (3 != max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV4), IV4 != 3]"),
-                 (3 > max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV6), IV6 < 3]"),
-                 (3 <= max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV8), IV8 >= 3]"),
-                ]
+@pytest.fixture()
+def setup():
+    _IntVarImpl.counter = 0
+    _BoolVarImpl.counter = 0
+    yield
 
-        # test transformation
-        for (expr, strexpr) in cases:
-            self.assertEqual( str(only_numexpr_equality(flatten_constraint(expr))), strexpr )
-            self.assertTrue(Model(expr).solve())
+class TestTransform:
+    @pytest.mark.parametrize(("constraint", "expected"),_generate_inputs(), ids=str)
+    def test_transforms(self, constraint, expected, setup):
```

Comment:
My concern about separating the logic from the test, this is of course just a simple testfile, with just one class whose purpose is to serve one function, which uses a paramtrize and a setup function; that is still overviewable, but if we would start applying this to all transform tests, I fear it would quickly become unoverseeable and unmaintainable...

Reply:
See previous two comments. We'd only need one class to run all tests of type after adding the transform to parametrize, given us test cases that check: "Given transform `f`, check that for input model `m` and expected output model `o`, we have `f(m)==o`. 

I don't think this approach will fit with every possible way you want to check a transform, but in many cases it will fit (and become more maintainable because you're not repeating the checking code over and over).